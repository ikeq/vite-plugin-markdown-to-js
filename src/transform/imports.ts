import camelCase from 'lodash/camelCase';
import { LangTransform } from '../parse';

const rSplit = /[\n\s,]/;
const rExt = /\.\w+$/;

/**
 * @example
 * ```md
 * ```imports
 * module1, module2, module3
 * module4
 * ```
 * ```
 */
export function transformImports({
  base,
  defaultPrefix,
}: { base?: string | ((path: string) => string); defaultPrefix?: string } = {}): LangTransform {
  const baseFn = (() => {
    if (typeof base === 'function') return base;
    if (typeof base === 'string') {
      if (base && !base.endsWith('/')) {
        base = `${base}/`;
      }
      return (path: string) => `${base || ''}${path}`;
    }
    return (path: string) => path;
  })();

  return {
    lang: 'imports',
    transform(src) {
      const imports = [...new Set(src.split(rSplit).filter((i) => i))];
      const importNames = imports.map((i) =>
        defaultPrefix ? camelCase(`${defaultPrefix} ${i}`) : i
      );

      return [
        {
          type: 'imports',
          imports: importNames,
          code: '',
        },
        {
          type: 'script',
          imports: importNames,
          code: imports
            .map((i, index) => {
              const path = baseFn(`${i}${rExt.test(i) ? '' : '.md'}`);
              return `import ${importNames[index]} from "${path}";`;
            })
            .join('\n'),
        },
      ];
    },
  };
}
