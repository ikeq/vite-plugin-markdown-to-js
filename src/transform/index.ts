export * from './imports';
export * from './vue';
