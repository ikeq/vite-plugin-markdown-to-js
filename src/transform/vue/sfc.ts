const tagOpeningRe = /\<((?:[a-z]+)(?:-[a-z]+)?)\s?/;

function parseTag(content: string, tag: string) {
  const matcher = {
    openingRe: new RegExp(`\\<${tag}([\\s\\S]*?)\\>`),
    closingRe: new RegExp(`\\<\\/${tag}\\>`)
  };
  let remaining = content;
  let pointer = -1;
  let indics = [];
  let attrs = '';

  while (remaining) {
    const opening = matcher.openingRe.exec(remaining);
    // none matching of open tag
    if (pointer < 0) {
      // malformed
      if (!opening) return null;
      // opening
      pointer += 1;
      indics[0] = opening.index + opening[0].length;
      attrs = opening[1];

      remaining = remaining.slice(opening.index + opening[0].length);
      continue;
    }

    // nesting
    if (opening && !matcher.closingRe.test(remaining.slice(0, opening.index))) {
      pointer += 1;

      remaining = remaining.slice(opening.index + opening[0].length);
      continue;
    } else {
      const closing = matcher.closingRe.exec(remaining);
      // malformed
      if (!closing) return null;

      pointer -= 1;
      remaining = remaining.slice(closing.index + closing[0].length);

      // closing
      if (pointer === -1) {
        indics[1] = content.length - remaining.length - closing[0].length;
        break;
      }
    }
  }

  return {
    code: content.slice(...indics),
    tag,
    attrs,
    remaining
  };
}

export function parseSFC<T>(raw: string): { code: string; tag: string & T; attrs: string }[] {
  const fn = (raw: string, ret: any = []) => {
    const group = tagOpeningRe.exec(raw);
    if (!group || !group[1]) return [];

    const block = parseTag(raw, group[1] as any);
    if (block) {
      ret.push({
        tag: block.tag,
        code: block.code,
        attrs: block.attrs
      });
      if (block.remaining) {
        fn(block.remaining, ret);
      }
    }

    return ret;
  }
  const blocks = fn(raw) as ReturnType<typeof parseSFC>;

  // merging
  const ret: Record<string, any> = {};

  blocks.forEach((i) => {
    if (!ret[i.tag]) {
      const attrs = i.attrs.trim();
      ret[i.tag] = {
        tag: i.tag,
        attrs: attrs ? ` ${attrs}` : '',
        code: ''
      };
    }
    ret[i.tag].code = ret[i.tag].code + '\n' + i.code.trim();
  });

  return Object.values(ret);
}
