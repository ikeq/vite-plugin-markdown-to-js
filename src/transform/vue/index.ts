import kebabCase from 'lodash/kebabCase';
import { CodeBlock, LangTransform } from '../../parse';
import { parseSFC } from './sfc';


const defaultRenders: Record<'template' | 'script' | 'style', (code: string, meta: any, env: Parameters<LangTransform['transform']>[2]) => string> = {
  template: code => code.trim(),
  style: code => code.trim(),
  script: (code, { imports }) => {
    code = code.trim();
    if (!code) {
      return `export default { components: { ${imports || ''} } };`;
    }
    return code.replace('export default {', `export default {\n components: { ${imports || ''} },`);
  },
};

/**
 * @example
 * ```md
 * ```vue
 * <template>
 * </template>
 * <script>
 * export default {}
 * </script>
 * ```
 * ```
 */
export function transformVue(options: {
  importsAsComponents?: boolean | ((imports: string[]) => string);
  renders?: typeof defaultRenders
} = {}): LangTransform {
  return {
    lang: 'vue',
    tokenize(tokens) {
      if (!tokens.some((token) => token.type === 'code' && token.lang === 'vue')) {
        tokens.push({ type: 'code', lang: 'vue', text: '<script></script>', raw: '' });
      }

      return tokens;
    },
    transform(src, md, env) {
      const renders: typeof options.renders = {
        template: options.renders?.template || defaultRenders.template,
        style: options.renders?.style || defaultRenders.style,
        script: options.renders?.script || defaultRenders.script,
      };
      const blocks = parseSFC<'template' | 'script' | 'style'>(src);
      const ret: CodeBlock[] = [{
        type: 'html',
        code: md.replace('```vue', '```html'),
        transform: source => `<!-- langvue -->${source.replace(/^<pre>/, '<pre v-pre>')}<!-- endlangvue -->`,
      }];

      if (!blocks.length) return ret;

      const tags = blocks.reduce((map, i) => {
        return {
          ...map,
          [i.tag]: { ...i }
        };
      }, {} as Record<typeof blocks[0]['tag'], typeof blocks[0]>);

      if (!tags.script) {
        tags.script = {
          tag: 'script',
          code: '',
          attrs: '',
        };
      }
      if (!tags.template) {
        tags.template = {
          tag: 'template',
          code: '',
          attrs: '',
        };
      }
      if (!tags.style) {
        tags.style = {
          tag: 'style',
          code: '',
          attrs: '',
        };
      }

      Object.keys(tags).forEach((key) => {
        const tagName = key as keyof typeof tags;
        const block = tags[tagName];
        const renderer = renders[tagName];
        const code: CodeBlock['code'] = tagName === 'script' && options.importsAsComponents
          ? (scripts, html) => {
            const imports = scripts.map((i) => i.imports || []).flat(2);

            html.forEach((h) => {
              // consume imports block
              if (h.type === 'imports' && h.imports) {
                h.type = 'html';
                h.skipCompile = true;
                h.code =
                  typeof options.importsAsComponents === 'function'
                    ? options.importsAsComponents(h.imports)
                    : (h.imports.map((i: string) => `<${kebabCase(i)}/>`).join('') as string);
              }
            });

            return renderer(block.code, { imports }, env)
          }
          : renderer(block.code, {}, env);

        ret.unshift({ type: tagName, attrs: block.attrs, code, skipCompile: true });
      });

      return ret;
    },
  };
}
