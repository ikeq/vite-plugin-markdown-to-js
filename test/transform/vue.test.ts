import { parseSFC } from '../../src/transform/vue/sfc';
import { transformVue } from '../../src/transform/vue';

const sfc = `
<template>
<div>
<template>asd</template>
</div>
</template>
<script>
export default { name: 'test' }
</script>
<style>
.test { color: red; }
</style>`;

const sfcWithAttrs = `
<template lang="html">
<div>
  <template
  a="1">asd</template>
</div>
</template>
<script lang="ts">
export default { name: 'test' }
</script>
<style lang="less" scoped>
.test { color: red; }
</style>`;

const sfcTemplateNesting = `
<template>
<div>
<template>
  <template>
    <template>asd</template>
  </template>
  <template>
    <template>asd</template>
  </template>
</template>
<template>
  <template>
    <template>asd</template>
  </template>
  <template>
    <template>asd</template>
  </template>
</template>
</div>
</template>
<style>
.test { color: red; }
</style>`;

const sfcMerging = `
<template foo>
<template>1</template>
</template>
<template bar>
<template>2</template>
</template>
<script lang="ts">
export const a = 1;
</script>
<script lang="js">
export default { name: 'test' };
</script>
<style lang="scss">
.test { color: red; }
</style>
<style lang="less">
.test { color: blue; }
</style>`;

const sfcSansScript = `
<template>
<div>
  <template>asd</template>
</div>
</template>
<style>
.test { color: red; }
</style>`;

describe('parseSFC', () => {
  it('basic', () => {
    expect(parseSFC(sfc)).toMatchSnapshot();
  });

  it('with attrs', () => {
    expect(parseSFC(sfcWithAttrs)).toMatchSnapshot();
  });

  it('template nesting', () => {
    expect(parseSFC(sfcTemplateNesting)).toMatchSnapshot();
  });

  it('merging', () => {
    expect(parseSFC(sfcMerging)).toMatchSnapshot();
  });
});

describe('transformVue', () => {
  it('basic', () => {
    const { transform } = transformVue();

    expect(transform(sfcMerging, '```vue\n```')).toMatchSnapshot();
  });

  it('sans script', () => {
    const { transform } = transformVue();

    expect(transform(sfcSansScript, '```vue\n```')).toMatchSnapshot();
  });

  it('importsAsComponents: true', () => {
    const { transform } = transformVue({ importsAsComponents: true });

    expect(transform(sfcMerging, '```vue\n```')).toMatchSnapshot();
  });

  it('importsAsComponents: Function', () => {
    const { transform } = transformVue({
      importsAsComponents(a) {
        return 'a'
      }
    });

    expect(transform(sfcMerging, '```vue\n```')).toMatchSnapshot();
  });
});
