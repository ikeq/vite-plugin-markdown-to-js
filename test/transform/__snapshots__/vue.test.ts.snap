// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseSFC basic 1`] = `
Array [
  Object {
    "attrs": "",
    "code": "
<div>
<template>asd</template>
</div>",
    "tag": "template",
  },
  Object {
    "attrs": "",
    "code": "
export default { name: 'test' }",
    "tag": "script",
  },
  Object {
    "attrs": "",
    "code": "
.test { color: red; }",
    "tag": "style",
  },
]
`;

exports[`parseSFC merging 1`] = `
Array [
  Object {
    "attrs": " foo",
    "code": "
<template>1</template>
<template>2</template>",
    "tag": "template",
  },
  Object {
    "attrs": " lang=\\"ts\\"",
    "code": "
export const a = 1;
export default { name: 'test' };",
    "tag": "script",
  },
  Object {
    "attrs": " lang=\\"scss\\"",
    "code": "
.test { color: red; }
.test { color: blue; }",
    "tag": "style",
  },
]
`;

exports[`parseSFC template nesting 1`] = `
Array [
  Object {
    "attrs": "",
    "code": "
<div>
<template>
  <template>
    <template>asd</template>
  </template>
  <template>
    <template>asd</template>
  </template>
</template>
<template>
  <template>
    <template>asd</template>
  </template>
  <template>
    <template>asd</template>
  </template>
</template>
</div>",
    "tag": "template",
  },
  Object {
    "attrs": "",
    "code": "
.test { color: red; }",
    "tag": "style",
  },
]
`;

exports[`parseSFC with attrs 1`] = `
Array [
  Object {
    "attrs": " lang=\\"html\\"",
    "code": "
<div>
  <template>asd</template>
</div>",
    "tag": "template",
  },
  Object {
    "attrs": " lang=\\"ts\\"",
    "code": "
export default { name: 'test' }",
    "tag": "script",
  },
  Object {
    "attrs": " lang=\\"less\\" scoped",
    "code": "
.test { color: red; }",
    "tag": "style",
  },
]
`;

exports[`transformVue basic 1`] = `
Array [
  Object {
    "attrs": " lang=\\"scss\\"",
    "code": "
.test { color: red; }
.test { color: blue; }",
    "skipCompile": true,
    "type": "style",
  },
  Object {
    "attrs": " lang=\\"ts\\"",
    "code": "
export const a = 1;
export default { name: 'test' };",
    "skipCompile": true,
    "type": "script",
  },
  Object {
    "attrs": " foo",
    "code": "
<template>1</template>
<template>2</template>",
    "skipCompile": true,
    "type": "template",
  },
  Object {
    "code": "\`\`\`html
\`\`\`",
    "type": "html",
  },
]
`;

exports[`transformVue importsAsComponents: Function 1`] = `
Array [
  Object {
    "attrs": " lang=\\"scss\\"",
    "code": "
.test { color: red; }
.test { color: blue; }",
    "skipCompile": true,
    "type": "style",
  },
  Object {
    "attrs": " lang=\\"ts\\"",
    "code": [Function],
    "skipCompile": true,
    "type": "script",
  },
  Object {
    "attrs": " foo",
    "code": "
<template>1</template>
<template>2</template>",
    "skipCompile": true,
    "type": "template",
  },
  Object {
    "code": "\`\`\`html
\`\`\`",
    "type": "html",
  },
]
`;

exports[`transformVue importsAsComponents: true 1`] = `
Array [
  Object {
    "attrs": " lang=\\"scss\\"",
    "code": "
.test { color: red; }
.test { color: blue; }",
    "skipCompile": true,
    "type": "style",
  },
  Object {
    "attrs": " lang=\\"ts\\"",
    "code": [Function],
    "skipCompile": true,
    "type": "script",
  },
  Object {
    "attrs": " foo",
    "code": "
<template>1</template>
<template>2</template>",
    "skipCompile": true,
    "type": "template",
  },
  Object {
    "code": "\`\`\`html
\`\`\`",
    "type": "html",
  },
]
`;

exports[`transformVue sans script 1`] = `
Array [
  Object {
    "attrs": "",
    "code": "export default { };",
    "skipCompile": true,
    "type": "script",
  },
  Object {
    "attrs": "",
    "code": "
.test { color: red; }",
    "skipCompile": true,
    "type": "style",
  },
  Object {
    "attrs": "",
    "code": "
<div>
  <template>asd</template>
</div>",
    "skipCompile": true,
    "type": "template",
  },
  Object {
    "code": "\`\`\`html
\`\`\`",
    "type": "html",
  },
]
`;
